odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2248 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Implementation of tensor spaces using ``pygpu``."""
+
+from __future__ import print_function, division, absolute_import
+import numpy as np
+
+from odl.set import RealNumbers
+from odl.space.base_tensors import TensorSpace, Tensor
+from odl.space.weighting import (
+    Weighting, ArrayWeighting, ConstWeighting,
+    CustomInner, CustomNorm, CustomDist)
+from odl.util import dtype_str, is_floating_dtype, signature_string
+
+try:
+    import cupy
+except ImportError:
+    CUPY_AVAILABLE = False
+else:
+    from pkg_resources import parse_version
+    if parse_version(cupy.__version__) < parse_version('2.0.0rc1'):
+        raise ImportError('cupy <2.0.0rc1 not supported')
+    CUPY_AVAILABLE = True
+
+
+__all__ = ('CupyTensorSpace',)
+
+
+# --- Space method implementations --- #
+
+
+lico = cupy.ElementwiseKernel(in_params='T a, T x, T b, T y',
```

Comment:
They should be, since `T` is a type template, so the only alternative would be to compile for all types, which I cannot imagine. The good thing is that the kernels are cached on disk, so after the first call things run way faster.

Reply:
I just checked. When I removed `pkg_resources` (bloody slow, we should avoid it whenever possible, at least during `import odl`), the import time of `cupy_tensors.py` is practically identical to the time of `import cupy` (~40 ms).